epoch_broadcaster/src/server.rs

```diff
@@ -25,40 +33,175 @@ impl EpochBroadcaster for ProtoServer {
         &self,
         request: Request<SetEpochRequest>,
     ) -> Result<Response<SetEpochResponse>, TStatus> {
+        let reply = SetEpochResponse {};
         let current_epoch = self.server.epoch.load(SeqCst);
```

Comment:
The `Ordering` parameters look too strict here and below. IIUC, since the only mutable state is the `AtomicUsize`, the `relaxed` ordering should be sufficient. I find [this](https://stackoverflow.com/questions/12346487/what-do-each-memory-order-mean/70585811#70585811) to be a fairly readable explanation.

Reply:
I agree that sequential consistency is probably too strong, but I am not sure that relaxed is strong enough. It is true that all threads will observe the same order of writes, but it doesn't seem like it's guaranteed that once the writer thread completes a write, subsequent reads from the reader threads will always see the value:
"It's as if relaxed operations propagate between threads with slight unpredictable delays."
 We need this before we can acknowledge the operation to the epoch service.

In any case, I think it's best to start with the safest semantics and only relax when we need to for performance. WDYT?

Reply:
> it's guaranteed that once the writer thread completes a write, subsequent reads from the reader threads
I read a bit more on this [1]. I think even SeqCst does not guarantee what I think you are shooting for. Basically, we need to guarantee that after we have responded to a `set_epoch` call, all chronologically subsequent `read_epoch` calls will return the new epoch. But CPUs don't define `subsequent` that way. They define it more of a database-y serialization way. From that article above, 
> What does “subsequent” mean? Java defines that all lock, unlock, and volatile variable accesses behave as if they ocurred in some sequentially consistent interleaving, giving a total order over all those operations in the entire program. “Subsequent” means later in that total order.

So compiler is free to continue reading and returning the stale value for an arbitrary amount of time from other threads - it just isn't allowed to read the new value and go back to returning old values. The different memory orderings does not change this behavior - rather they deal with the order of how combinations of atomic and non-atomic variable modifications become visible in different threads. I think to get that behavior we want we'd need to establish a happens-before relationship between writers and readers i.e. `set_epoch` and `read_epoch`. The simplest approach would be some sort of reader-writer lock where we block writes behind ongoing reads and new reads wait for writes to finish.

[1] https://research.swtch.com/plmm

Reply:
This is also a good read I think: https://doc.rust-lang.org/nomicon/atomics.html#acquire-release
Additionally, I think Acquire/Release is what we really want here since we are only dealing with 1 variable.

Reply:
I am confused about what `previous` means here. Is it a temporal previous or causal previous? Also from `fence` documentation [1], 

> A fence ‘A’ which has (at least) [Release](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Release) ordering semantics, synchronizes with a fence ‘B’ with (at least) [Acquire](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Acquire) semantics, if and only if there exist operations X and Y, both operating on some atomic object ‘M’ such that A is sequenced before X, Y is sequenced before B and Y observes the change to M. This provides a happens-before dependence between A and B.

Note the `Y observes the change to M` part. It says nothing about passage of time.

I agree that acquire-release will be just as good as SeqCst for this case.

[1] https://doc.rust-lang.org/std/sync/atomic/fn.fence.html